services/src/main/java/org/keycloak/broker/saml/mappers/XPathAttributeMapper.java

```diff
@@ -0,0 +1,316 @@
+        };
+    }
+
+    private static Function<String, Object> applyXPath(String attributeXPath) {
+        return xml -> {
+            try {
+                LOGGER.trace("Trying to parse: " + xml);
+
+                Matcher namespaceMatcher = Pattern.compile("xmlns:(\\w+)=\"(.+?)\"").matcher(xml);
+                Map<String, String> namespaces = new HashMap<>();
+                Map<String, String> prefixes = new HashMap<>();
+                while (namespaceMatcher.find()) {
+                    namespaces.put(namespaceMatcher.group(1), namespaceMatcher.group(2));
+                    prefixes.put(namespaceMatcher.group(2), namespaceMatcher.group(1));
+                }
+
+                XPath xPath = XPathFactory.newInstance().newXPath();
+                xPath.setNamespaceContext(new NamespaceContext() {
+                    @Override
+                    public String getNamespaceURI(String prefix) {
+                        if (namespaces.containsKey(prefix)) {
+                            return namespaces.get(prefix);
+                        }
+
+                        return XMLConstants.NULL_NS_URI;
+                    }
+
+                    @Override
+                    public String getPrefix(String namespaceURI) {
+                        if (prefixes.containsKey(namespaceURI)) {
+                            return prefixes.get(namespaceURI);
+                        }
+
+                        return null;
+                    }
+
+                    @Override
+                    public Iterator<String> getPrefixes(String namespaceURI) {
+                        List<String> list = new ArrayList<>();
+                        if (prefixes.containsKey(namespaceURI)) {
+                            list.add(prefixes.get(namespaceURI));
+                        }
+
+                        return list.iterator();
+                    }
+                });
+                Document document = DocumentUtil.getDocument(new StringReader(xml));
+                return xPath.compile(attributeXPath).evaluate(document, XPathConstants.STRING);
+            } catch (XPathExpressionException e) {
+                LOGGER.warn("Unparsable element will be ignored", e);
+                return "";
+            } catch (Exception e) {
+                throw new RuntimeException("Could not parse xml element", e);
+            }
+        };
+    }
+
+    private List<String> findAttributeValuesInContext(String attributeName, String attributeXPath, BrokeredIdentityContext context) {
+        AssertionType assertion = (AssertionType) context.getContextData().get(SAMLEndpoint.SAML_ASSERTION);
+
+        return assertion.getAttributeStatements().stream()
+                .flatMap(statement -> statement.getAttributes().stream())
+                .filter(elementWith(attributeName))
+                .flatMap(attributeType -> attributeType.getAttribute().getAttributeValue().stream())
```

Comment:
Consider leveraging method references rather than introducing new methods. https://softwareengineering.stackexchange.com/a/285311 can give more insight why:

```suggestion
                .map(AttributeStatementType::getAttributes)
                .flatMap(Collection::stream)
                .map(List::getAttributes)
                .filter(elementWith(attributeName))
                .map(AttributeStatementType.ASTChoiceType::getAttribute)
                .filter(Objects::nonNull) // note that the attribute could be encrypted, thus resulting to null
                .map(AttributeType::getAttributeValue)
                .flatMap(Collection::stream)
```